### PR TITLE
impl fallible IntoIterator for ListModel,ListStore

### PR DESCRIPTION
**Pull request by ranfdev**
It's me with another take on #204.

### CODE REVIEW

**Comment by ranfdev** [comment]
My only question is: once the error `ListModelMutatedDuringIter` is returned, when should the iterator stop? Should we keep returning `Some(Err(...))` for the remaining items (`initial_model_size - current_i`), ad infinitum, or just one time?

File: gio/src/list_model.rs [hunk 1]
```diff
@@ -19,15 +30,86 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+}
+impl<T: IsA<glib::Object>> Iterator for ListModelIter<T> {
+    type Item = Result<T, ListModelMutatedDuringIter>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let res = match self.changed.get() {
+            true => Some(Err(ListModelMutatedDuringIter)),
+            false => self
+                .model
+                .item(self.i)
```
Comment (sdroege) [inline gio/src/list_model.rs:76]: This can't possibly return `None` here so I would `unwrap()` / `expect()` right after `.item()` the first time

File: gio/src/list_model.rs [hunk 2]
```diff
@@ -19,15 +30,86 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+                .map(|x| Ok(x.downcast::<T>().unwrap())),
```
Comment (sdroege) [inline gio/src/list_model.rs:77]: ... and then a second time after `downcast()` because that can't really fail either

File: gio/src/list_model.rs [hunk 3]
```diff
@@ -19,15 +30,86 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+    i: u32,
+    model: ListModel,
+    changed: Rc<Cell<bool>>,
+    signal_id: Cell<Option<SignalHandlerId>>,
+}
+impl<T: IsA<glib::Object>> Iterator for ListModelIter<T> {
+    type Item = Result<T, ListModelMutatedDuringIter>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let res = match self.changed.get() {
```
Comment (sdroege) [inline gio/src/list_model.rs:72]: You should get the number of items in the model when creating the iterator and then work based on that. The number of items can't change duration iteration, and by doing that you can implement
  - `Iterator::size_hint()` and `ExactSizeIterator`
  - `FusedIterator`
  - `DoubleEndedIterator`

File: gio/src/list_model.rs [hunk 4]
```diff
@@ -19,15 +30,86 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
         }
```
Comment (sdroege) [inline gio/src/list_model.rs:30]: You can implement ^---- this one via `self.iter().collect::<Result<Vec<T>, _>>>().unwrap()` now

File: gio/src/list_model.rs [hunk 5]
```diff
@@ -19,15 +30,86 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
         res
     }
+    fn iter<LT: IsA<glib::Object>>(&self) -> Result<ListModelIter<LT>, &Self>
+    where
+        Self: Sized + Clone,
+    {
+        if self.item_type() != LT::static_type() {
```
Comment (sdroege) [inline gio/src/list_model.rs:37]:
```suggestion
        if !self.item_type().is_a(LT::static_type()) {
```

File: gio/src/list_model.rs [hunk 6]
```diff
@@ -19,15 +30,86 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+
+        let rcc = rc.clone();
+        let signal_id = Cell::new(Some(self.connect_items_changed(move |_, _, _, _| {
+            rcc.replace(true);
+        })));
+
+        Ok(ListModelIter {
+            ty: Default::default(),
+            i: 0,
+            model: self.clone().upcast(),
```
Comment (sdroege) [inline gio/src/list_model.rs:50]: Why not just store a reference in here instead of cloning?

File: gio/src/list_model.rs [hunk 7]
```diff
@@ -19,15 +30,86 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+        res
+    }
+}
+impl<T: IsA<glib::Object>> Drop for ListModelIter<T> {
+    fn drop(&mut self) {
+        self.model.disconnect(self.signal_id.take().unwrap());
+    }
 }
 
 impl std::iter::IntoIterator for ListModel {
```
Comment (sdroege) [inline gio/src/list_model.rs:89]:
```suggestion
impl std::iter::IntoIterator for &ListModel {
```

makes more sense as you don't actually consume the model (compare: `IntoIterator` on `Vec` and `&Vec`).

File: glib/src/object.rs [hunk 8]
```diff
@@ -2263,16 +2263,13 @@ impl<T: ObjectType> ObjectExt for T {
...
-            Some(pspec) => pspec,
-            None => {
-                return Err(bool_error!(
-                    "property '{}' of type '{}' not found",
-                    property_name,
-                    self.type_()
-                ));
-            }
-        };
+        let pspec = self.find_property(property_name).ok_or_else(|| {
```
Comment (sdroege) [inline glib/src/object.rs:2266]: Can you put all this cleanup into a separate PR? That's going to be merged faster and keeps this PR smaller then :)

**Review by sdroege — COMMENTED** [review]
The general idea of having an iterator that returns a `Result` seems good to me. That seems like the simplest and most sensible implementation for how people would actually use this.

File: gio/src/list_model.rs (hunk 6 above)
Comment (ranfdev) [reply gio/src/list_model.rs:50]: because `upcast` consumes `self`:
`pub fn upcast<T>(self)`

File: gio/src/list_model.rs (hunk 4 above)
Comment (ranfdev) [reply gio/src/list_model.rs:30]: I'd keep `snapshot` as it is, because now it's faster and the implementation is really small anyway

File: gio/src/list_model.rs (hunk 6 above)
Comment (sdroege) [reply gio/src/list_model.rs:54]: You could do `model: self.upcast_ref(),` and just store a reference. That also makes it less strange as iterators usually keep a normal reference to their underlying data

File: gio/src/list_model.rs [hunk 9]
```diff
@@ -19,15 +31,149 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
         }
         res
     }
+    fn iter<LT: IsA<glib::Object>>(&self) -> Result<ListModelIter<LT>, &Self>
+    where
+        Self: Sized + Clone,
+    {
+        if !self.item_type().is_a(LT::static_type()) {
+            return Err(self);
```
Comment (sdroege) [inline gio/src/list_model.rs:39]: Would be nicer to return a proper error type here. Returning `Self` only really makes sense for functions that take it by value

File: gio/src/list_model.rs [hunk 10]
```diff
@@ -19,15 +31,149 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+        let n = (self.reverse_pos - self.i) as usize;
+        (n as usize, Some(n))
+    }
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.i >= self.reverse_pos {
+            return None;
+        }
+        let res = match self.changed.get() {
+            true => Err(ListModelMutatedDuringIter),
```
Comment (sdroege) [inline gio/src/list_model.rs:93]: You probably want to return `None` forever after returning `Err` once

File: gio/src/list_model.rs [hunk 11]
```diff
@@ -19,15 +31,149 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+/// will return `Some(Err(...))` for the remaining items.
+/// Mutations to the `ListModel` in position >= `initial_model.n_items()` are allowed.
+pub struct ListModelIter<T: IsA<glib::Object>> {
+    ty: PhantomData<T>,
+    i: u32,
+    // it's > i when valid
+    reverse_pos: u32,
+    model: ListModel,
+    changed: Rc<Cell<bool>>,
+    signal_id: Cell<Option<SignalHandlerId>>,
```
Comment (sdroege) [inline gio/src/list_model.rs:79]:
```suggestion
    signal_id: Option<SignalHandlerId>,
```

Doesn't have to be a cell. You only mutate it when you have a mutable reference anyway

File: gio/src/list_model.rs [hunk 12]
```diff
@@ -19,15 +31,149 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+        if !self.item_type().is_a(LT::static_type()) {
+            return Err(self);
+        }
+
+        let len = self.n_items();
+        let changed = Rc::new(Cell::new(false));
+
+        let changed_clone = changed.clone();
+        let signal_id = Cell::new(Some(self.connect_items_changed(move |_, pos, _, _| {
+            changed_clone.replace(pos < len);
```
Comment (sdroege) [inline gio/src/list_model.rs:47]:
```suggestion
            let old = changed_clone.get();
            changed_clone.set(old || pos < len);
```

File: gio/src/list_model.rs [hunk 13]
```diff
@@ -19,15 +31,149 @@ impl<T: IsA<ListModel>> ListModelExtManual for T {
...
+}
+impl<T: IsA<glib::Object>> Iterator for ListModelIter<T> {
+    type Item = Result<T, ListModelMutatedDuringIter>;
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let n = (self.reverse_pos - self.i) as usize;
+        (n as usize, Some(n))
+    }
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.i >= self.reverse_pos {
```
[the remark this answers is not in the document]
Comment (ranfdev) [reply gio/src/list_model.rs:87]: They both work equally well. In this case I read `==` as "If both are equal, return early. The code below shouldn't be executed".
I prefer `>=` because it also says "The code below should be executed ONLY if `self.i < self.reverse_pos`".

But I see why you may like `==`: "`self.i` can't never be greater than `self.reverse_pos`, so why check?"

Whatever... :)

File: gio/src/list_model.rs (hunk 10 above)
Comment (ranfdev) [reply gio/src/list_model.rs:91]: Wouldn't this break the trait `ExactIterator`? I can't return `None` earlier than expected. I'm ok with returning `None`, but now while `ExactIterator` is implemented

File: gio/src/list_model.rs (hunk 9 above)
Comment (ranfdev) [reply gio/src/list_model.rs:36]: True, but this keeps `iter` similar to `downcast`. This type check feels similar to what downcast is doing. I know downcast returns `Err(Self)`... I try to return the most similar thing I can

File: gio/src/list_model.rs (hunk 10 above)
Comment (sdroege) [reply gio/src/list_model.rs:91]: That's a good point, yes you're right :)

File: gio/src/list_model.rs (hunk 9 above)
Comment (sdroege) [reply gio/src/list_model.rs:36]: `downcast_ref()` returns an `Option` :) The only reason why `downcast()` returns `Self` in the error case is because it takes `Self` by value and it would otherwise be gone on error.

Comment (ranfdev) [reply gio/src/list_model.rs:36]: > The only reason why downcast() returns Self in the error case is because it takes Self by value and it would otherwise be gone on error.

I hadn't thought of that 😅

**Review by sdroege — APPROVED** [review]
